### DISCUSSION TITLE
multi-threaded uploading/downloading (Issue#232)

### DISCUSSION THREAD

**Discussion by npapatheodorou** [abraunegg/onedrive#2387] [General]
Dear abraunegg,

I am a regular user of your OneDrive client and I was wondering if there have been any updates regarding the multi-threaded uploading and downloading feature request that was discussed earlier. As a user who often works with large files, I find that the current single-threaded transfer process can be slow and inefficient, and I'm hoping that the addition of multi-threading capabilities could significantly improve the uploading and downloading experience on OneDrive.

I was wondering if you could provide an update on the progress of the multi-threaded uploading and downloading feature request. Is the development still ongoing, and if so, is there any estimated timeline for when this feature might be implemented? As a user, I am eager to see this feature become available, and any information you can provide on the status of this request would be greatly appreciated.

Thank you very much for your time and attention. I appreciate any information you can provide on this matter.

Sincerely,
Nikos

**Comment by abraunegg** [comment]
Specifically to answer this question:

> As a user who often works with large files, I find that the current single-threaded transfer process can be slow and inefficient, and I'm hoping that the addition of multi-threading capabilities could significantly improve the uploading and downloading experience on OneDrive.

Multi-threading is not going to solve your slow transfer process. Each 'file' will be uploaded|downloaded in a new thread. So instead of 1 file being done at a time before proceeding to the next, _X_ number files will be transferred at a time. The time to transfer each file will not change. At the moment I am trying to determine the optimum X value (see below).

If you are thinking that this feature is all about splitting up a large file in multiple chunks, and uploading|downloading each chunk in a separate thread - the OneDrive API has zero capability for this - so this is not the feature that is going to solve your problem in that regard.

To all watching for this Feature Request

> I was wondering if you could provide an update on the progress of the multi-threaded uploading and downloading feature request

The high level framework & architecture design is done, and I consider this a prototype at the moment.

### What does the current prototype do ?
The client connects to the OneDrive API, downloads all files online in a multi-threaded manner. I am in the process of tuning this to get the appropriate level so that the OneDrive API does not complain.

### What is missing ?
* Everything that v2.4.23 can do (except for downloading) - and even then, there are issues such as error handling, notifications, logging not even added in yet.

### Can I access it / test it ?
* Absolutely not. It is in no way shape or functional level that I would allow *anyone* to use this against their data, let alone test data. Many items in the code at the moment are *hard coded* to make it function - and this is 100% less than ideal.

### When may this be available ?
* As per the [release notes for v2.4.23](https://github.com/abraunegg/onedrive/releases/tag/v2.4.23)
![image](https://user-images.githubusercontent.com/4956234/232905880-41492ff7-d317-4f96-987e-ab23373d7e4b.png)

* As per the [feature request](https://github.com/abraunegg/onedrive/issues/232) 
![image](https://user-images.githubusercontent.com/4956234/232906275-41859293-105b-4d81-908a-47bbf16a7be0.png)

The re-write is being done for v2.5.x code branch and the plan is that https://github.com/abraunegg/onedrive/issues/232 is the first cab off the rank with v2.5.0.

A complication with this re-write however are the following 2 Feature Requests:
* https://github.com/abraunegg/onedrive/issues/1251
* https://github.com/abraunegg/onedrive/issues/1300

When bringing in feature parity with v2.4.23, the whole existing OneDrive Business Shared Folders feature will have to be 100% re-written, thus will be a heavily breaking change for *anyone* using OneDrive Business Shared Folders today. I need to think how to handle this gracefully.

The next aspect outside of that is *hopefully* (and given the architecture mapping that I have done) - the redesign will also allow https://github.com/abraunegg/onedrive/issues/757 to be developed in such a way that it will be possible to control this via a flag in the config file. It is too early to tell if this is workable, but the provision is in the design.

I am expecting at least another 3-6 months (maybe even the rest of the year due to time constraints) before even an 'Alpha 1' may be available for *anyone* to test.

When an 'Alpha 1' is available - I will be taking it very slow from that to even a Release Candidate - so that as many people can test the whole re-write as there will be bugs, there will be things that I cannot test for, there will be certain setups that folk are using that are 100% unique to them. 

To move through from Release Candidate -> Release I probably will even ask for users to submit test reports - and have at least 100-200 or more unique reports required *before* moving to the next phase to ensure that any issues, performance regressions and bugs are found before any final release of v2.5.0.
